src/jdk.compiler/share/classes/com/sun/tools/javac/code/TypeAnnotations.java

```diff
@@ -1423,6 +1425,12 @@ private void setTypeAnnotationPos(List<JCAnnotation> annotations, TypeAnnotation
             }
         }
 
+        private void checkArraySize(List<JCExpression> tree) {
+            //check if attribute length exceeds maximum unsigned 16-bit value
+            if (!sigOnly && tree != null && (tree.size() >>> 16) > 0) {
```

Comment:
Should we use `(tree.size & ~0xFFFF) != 0`? This is how classfile API checks the value compatibility, and I wonder how other places in javac does so.

Reply:
I was trying to chose between this and `tree.size & ~0xFFFF`, I think `javac` will sometimes check if an int is larger than `0xFFFF`

Reply:
If you use `& ~0xFFFF` you must use `!= 0` as the mask includes the sign bit.

Reply:
I think the other checks are implemented as simple `value > limit => error` checks:
https://github.com/openjdk/jdk/blob/d8430efb5e159b8e08d2cac66b46cb4ff1112927/src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Code.java#L114
https://github.com/openjdk/jdk/blob/d8430efb5e159b8e08d2cac66b46cb4ff1112927/src/jdk.compiler/share/classes/com/sun/tools/javac/jvm/Gen.java#L938

True, it may overflow, but if it would overflow once, it might overflow twice (back to positive numbers), so if/when that becomes a problem, we'd probably need something more complex. And it will be easier to find if the pattern is as close to the existing patterns as possible. So, I guess I would suggest to use a simple ` > ClassFile.MAX_ANNOTATIONS`.